File: wasm/verify/overlay_test.mjs
// core.ZStack against the minimal DOM.
//
// htmlout writes the same layout into a static document and is tested there;
// this is the live half, where the interesting properties are the two a static
// export cannot have — that a stack with no Style at all is still an overlay
// (createElement, not applyStyle, is what plants the chassis), and that a
// layer arriving in a *patch* is placed in the cell like the ones that were
// there at mount.
//
// The second is the one that would have shipped. An unplaced layer is not
// invisible: CSS auto-places it into its own implicit grid row, so it appears
// below the stack rather than on it, which reads as a layout quirk rather than
// as a missing declaration.

import test from "node:test";
import assert from "node:assert/strict";

import { loadRuntime, nodeAt } from "./load.mjs";

const text = (content) => ({ Type: "Text", Props: { content } });

// Mounts a ZStack as the root's only child and hands back the stack, a layer
// accessor, and the runtime for patching.
function mountStack(style, children) {
    const rt = loadRuntime();
    rt.GrMob.mount(JSON.stringify({
        Type: "Column",
        Children: [{ Type: "ZStack", Style: style, Children: children }],
    }));
    rt.drainFrames();
    return {
        rt,
        stack: nodeAt(rt.document, "root/0"),
        layer: (i) => nodeAt(rt.document, `root/0/${i}`),
    };
}

test("a stack is a single-cell grid and every layer is in the cell", () => {
    const { stack, layer } = mountStack({ Width: "160px" }, [text("under"), text("over")]);

    assert.equal(stack.style.display, "grid");
    // The *items* properties: they place a child inside its cell. The content
    // pair would place the single track inside the container, which on an
    // auto-sized container does nothing at all.
    assert.equal(stack.style.alignItems, "center");
    assert.equal(stack.style.justifyItems, "center");
    assert.equal(layer(0).style.gridArea, "1/1");
    assert.equal(layer(1).style.gridArea, "1/1");
});

test("a stack with no Style is still an overlay", () => {
    // core.ZStack carries no theme base, so this is the shape a stack written
    // with children and nothing else arrives in — and the one that never
    // reaches applyStyle at all.
    const { stack, layer } = mountStack(undefined, [text("under"), text("over")]);

    assert.equal(stack.style.display, "grid");
    assert.equal(stack.style.alignItems, "center");
    assert.equal(layer(0).style.gridArea, "1/1");
    assert.equal(layer(1).style.gridArea, "1/1");
});

test("a container prop does not promote a stack to a flex container", () => {
    // Gap is meaningless on an overlay — there is one cell and nothing to
    // space along — but the flex promotion test would have answered to it,
    // and the stack would have stopped overlaying entirely.
    const { stack } = mountStack({ Gap: 8, AlignItems: "flex-end" }, [text("a"), text("b")]);

    assert.equal(stack.style.display, "grid");
    assert.equal(stack.style.flexDirection, "");
    assert.equal(stack.style.alignItems, "center",
        "an AlignItems on a stack overrode the centring the alignment contract fixes");
});

test("an update-style patch does not cost a stack its grid", () => {
    // styleFromGrMob is total: it reassigns every property it manages on every
    // call, so the chassis planted at creation has to be restated there or the
    // first patch drops the stack into block flow.
    const { rt, stack } = mountStack({ Width: "160px" }, [text("under"), text("over")]);

    rt.GrMob.patch(JSON.stringify([{
        Type: "update-style", TargetID: "root/0", Changes: { Width: "200px" },
    }]));

    assert.equal(stack.style.width, "200px");
    assert.equal(stack.style.display, "grid");
    assert.equal(stack.style.justifyItems, "center");
});

test("a layer added by a patch is placed in the cell", () => {
    const { rt } = mountStack(undefined, [text("under")]);

    rt.GrMob.patch(JSON.stringify([
        { Type: "add-child", TargetID: "root/0", Changes: text("over") },
    ]));

    assert.equal(nodeAt(rt.document, "root/0/1").style.gridArea, "1/1",
        "the new layer was auto-placed into its own grid row, i.e. below the stack");
});

test("a replaced layer is placed in the cell", () => {
    // The patch names the layer, not the stack, so the post-batch pass has to
    // walk up to find the overlay above it — the same walk the TabView and
    // end-reached passes make.
    const { rt } = mountStack(undefined, [text("under"), text("over")]);

    rt.GrMob.patch(JSON.stringify([
        { Type: "replace", TargetID: "root/0/1", Changes: text("over (edited)") },
    ]));

    assert.equal(nodeAt(rt.document, "root/0/1").style.gridArea, "1/1");
});

test("an ordinary container does not place its children in a cell", () => {
    // The stamp is the overlay's, and a container that is not one must leave
    // its children alone — a grid-area on a flex item is inert, but a stray
    // one is the sign the pass is walking further than it should.
    const rt = loadRuntime();
    rt.GrMob.mount(JSON.stringify({
        Type: "Column",
        Children: [{ Type: "Box", Children: [text("a"), text("b")] }],
    }));
    rt.drainFrames();

    assert.equal(nodeAt(rt.document, "root/0/0").style.gridArea, undefined);
    assert.equal(nodeAt(rt.document, "root/0").style.display, "flex");
});

test("a layer names its own corner and the rest stay centred", () => {
    const { layer } = mountStack({ Width: "160px" }, [
        text("under"),
        { Type: "Text", Props: { content: "N" }, Style: { StackAlign: "top" } },
        { Type: "Text", Props: { content: "SE" }, Style: { StackAlign: "bottom-end" } },
    ]);

    // The unplaced layer is placed *explicitly* at the centre rather than left
    // to the chassis. See STACK_PLACEMENTS: the row exists so that a layer's
    // own AlignSelf cannot move it, and so that a layer losing its placement
    // comes back to the middle.
    assert.equal(layer(0).style.justifySelf, "center");
    assert.equal(layer(0).style.alignSelf, "center");

    assert.equal(layer(1).style.justifySelf, "center");
    assert.equal(layer(1).style.alignSelf, "start");

    assert.equal(layer(2).style.justifySelf, "end");
    assert.equal(layer(2).style.alignSelf, "end");
});

test("a layer's own AlignSelf does not move it", () => {
    // Style.AlignSelf is flexbox's, and a grid item honours it too — so
    // before the stack imposed a placement on every layer, this one prop
    // moved a layer on the two DOM targets and nowhere else, in contradiction
    // of the alignment contract core.ZStack documents.
    const { layer } = mountStack({ Width: "160px" }, [
        { Type: "Text", Props: { content: "a" }, Style: { AlignSelf: "flex-end" } },
    ]);

    assert.equal(layer(0).style.alignSelf, "center",
        "a flex item property placed a layer of an overlay");
});

test("a layer that loses its placement returns to the centre", () => {
    // The totality half. applyStyle drops the data attribute and the overlay
    // pass restates both properties, so nothing is left holding the old
    // corner — the failure a pass that only wrote placements it was asked for
    // would have.
    const { rt, layer } = mountStack({ Width: "160px" }, [
        { Type: "Text", Props: { content: "N" }, Style: { StackAlign: "top-start" } },
    ]);
    assert.equal(layer(0).style.justifySelf, "start");

    rt.GrMob.patch(JSON.stringify([{
        Type: "update-style", TargetID: "root/0/0", Changes: { FontSize: 12 },
    }]));

    assert.equal(layer(0).style.justifySelf, "center");
    assert.equal(layer(0).style.alignSelf, "center");
});

test("a placement outside an overlay is inert", () => {
    // The prop is imposed by the stack, so a node that is not a layer never
    // receives it — which is what keeps a StackAlign written on the wrong node
    // from re-placing a Row's children on the web alone.
    const rt = loadRuntime();
    rt.GrMob.mount(JSON.stringify({
        Type: "Column",
        Children: [{
            Type: "Row",
            Children: [{ Type: "Text", Props: { content: "a" }, Style: { StackAlign: "top-end" } }],
        }],
    }));
    rt.drainFrames();

    // Falsy rather than "": dom.mjs leaves a property that was never assigned
    // undefined, where a browser's CSSOM reports the empty string, and
    // justify-self is never assigned outside an overlay while align-self is
    // (styleFromGrMob writes it for every node, from Style.AlignSelf).
    const child = nodeAt(rt.document, "root/0/0");
    assert.ok(!child.style.justifySelf, "a StackAlign placed a node that is not a layer");
    assert.ok(!child.style.alignSelf, "a StackAlign placed a node that is not a layer");
});

File: wasm/verify/overlay_test.go
package main

import (
	"reflect"
	"regexp"
	"sort"
	"strings"
	"testing"

	"github.com/rohanthewiz/grmob/htmlout"
)

// The set of node types whose children are drawn on top of one another.
//
// Go is the authority (overlayTypes in htmlout/stack.go) and the runtime
// restates it, so the two are compared here rather than remembered — the same
// treatment the tag, input-type, generic-tag, stack-axis and border-reset
// tables get. Drift in either direction is silent and total: the layers of one
// overlay stack down the page on one web target and lie on top of each other
// on the other, and nothing errors on either side.
//
// Parsed with a regexp rather than through parseRuntimeTable, which reads
// key/value object literals; this one is a flat array of strings, as
// BORDER_RESET_TYPES is.
func TestRuntimeOverlayTypesMatchGo(t *testing.T) {
	src := runtimeSource(t)

	m := regexp.MustCompile(`const OVERLAY_TYPES = new Set\(\[([^\]]*)\]\);`).FindStringSubmatch(src)
	if m == nil {
		t.Fatalf("grmob-runtime.js: no `const OVERLAY_TYPES = new Set([...])` found — if it was " +
			"renamed or spread over several lines, update this test rather than deleting it")
	}
	var got []string
	for _, q := range regexp.MustCompile(`"([^"]*)"`).FindAllStringSubmatch(m[1], -1) {
		got = append(got, q[1])
	}
	sort.Strings(got)

	want := htmlout.OverlayTypes()
	if !reflect.DeepEqual(got, want) {
		t.Errorf("OVERLAY_TYPES is %v; htmlout.OverlayTypes() is %v — a node type overlays its "+
			"children on one web target and stacks them on the other", got, want)
	}
}

// The cell itself, which is the other half of the same contract and lives in a
// second constant because the two renderers write it in different places: an
// inline declaration list here, a CSSOM property there.
//
// Compared as the *value* rather than the whole declaration, since Go writes
// "grid-area:1/1" into a style attribute and the runtime assigns "1/1" to
// element.style.gridArea. A mismatch would put the layers of one stack in
// different cells on the two targets, which is a grid of layers on one and a
// pile on the other.
func TestRuntimeOverlayChildAreaMatchesGo(t *testing.T) {
	src := runtimeSource(t)

	m := regexp.MustCompile(`const OVERLAY_CHILD_AREA = "([^"]*)";`).FindStringSubmatch(src)
	if m == nil {
		t.Fatal("grmob-runtime.js: no `const OVERLAY_CHILD_AREA = \"...\"` found")
	}
	want := strings.TrimPrefix(htmlout.OverlayChildDecl, "grid-area:")
	if want == htmlout.OverlayChildDecl {
		t.Fatalf("htmlout.OverlayChildDecl is %q, which no longer names grid-area; this "+
			"comparison cannot strip a property it cannot find", htmlout.OverlayChildDecl)
	}
	if m[1] != want {
		t.Errorf("OVERLAY_CHILD_AREA is %q; htmlout places layers in %q", m[1], want)
	}
}

// The runtime must consult the set in styleFromGrMob's *first* branch, ahead
// of the flex test.
//
// Both facts matter and neither is visible from the table comparison above.
// Placing the overlay arm after the flex test would let a stack carrying a Gap
// or an AlignItems be promoted to a flex container and stop overlaying
// altogether — the loss of the whole node type, caused by a prop that is
// merely inert on an overlay. And an overlay branch that did not also clear
// flexDirection would break the function's totality rule, leaving the
// direction the box last had.
func TestRuntimeResolvesTheOverlayBeforeTheFlexTest(t *testing.T) {
	src := runtimeSource(t)

	overlay := strings.Index(src, `const overlay = OVERLAY_TYPES.has(nodeType);`)
	if overlay < 0 {
		t.Fatal("grmob-runtime.js: styleFromGrMob does not read OVERLAY_TYPES at all — a ZStack " +
			"renders as an ordinary block-flow div in the live app")
	}
	flex := strings.Index(src, `} else if (style.Gap || style.RowGap`)
	if flex < 0 {
		t.Fatal("grmob-runtime.js: the flex promotion test is no longer the arm after the " +
			"overlay one; update this test rather than deleting it")
	}
	if overlay > flex {
		t.Error("grmob-runtime.js: the overlay is resolved after the flex test, so a ZStack " +
			"carrying a Gap becomes a flex container and stops overlaying")
	}
	// The centring, which is the alignment contract core.ZStack documents, and
	// the two totality clears beside it.
	for _, expr := range []string{
		`out.alignItems = overlay ? "center" : (alignItems || "");`,
		`out.justifyItems = overlay ? "center" : "";`,
	} {
		if !strings.Contains(src, expr) {
			t.Errorf("grmob-runtime.js: styleFromGrMob is missing %q", expr)
		}
	}
}

// The layers are stamped by a pass over the container's children, and the pass
// has to run on the patch path too.
//
// The initial render is the easy half. An "add" patch lands a brand new layer
// under a ZStack that was not itself touched, so a runtime that stamped only
// at creation would place every layer of a stack except the ones that appeared
// later — and a layer with no cell is auto-placed into its own implicit grid
// row, i.e. below the stack instead of on it.
func TestRuntimeStampsOverlayLayersOnBothPaths(t *testing.T) {
	src := runtimeSource(t)
	for _, expr := range []string{
		"child.style.gridArea = OVERLAY_CHILD_AREA;", // the pass itself
		"syncOverlay(el);",                           // the initial render
		"syncTouchedOverlays(touched);",              // the patch path
	} {
		if !strings.Contains(src, expr) {
			t.Errorf("grmob-runtime.js: missing %q — an overlay's layers are not all placed "+
				"in its cell", expr)
		}
	}
}
